### ISSUE TITLE
Better support for the Parrot parser

### ISSUE THREAD

**Issue by paulk-asert** [CodeNarc/CodeNarc#414]
I found when using CodeNarc within the Groovy 3 build that some files were failing CodeNarc processing. There are very minor changes in how the new parser records line and column information to improve accuracy of that information and those changes seem to trip up processing in a few places. I'll do a bit more testing and then create a PR but the following diff seemed to fix most issues if anyone else wants to test in parallel. This seemed to fix most issues using Groovy 3 and didn't see to create new ones with 2.5. I did notice that I was getting worse false positives for rules being run on text within triple quoted strings. E.g. for a template generating Java source code it was telling me that the semicolons weren't needed. I'll investigate some more. The -2 to -1 change in the diff is perhaps coming into play there.

```
Index: src/main/groovy/org/codenarc/rule/unnecessary/UnnecessaryPublicModifierRule.groovy
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main/groovy/org/codenarc/rule/unnecessary/UnnecessaryPublicModifierRule.groovy	(revision 9a7cec02cb8cbaf845030f2434309e8968f5d7a7)
+++ src/main/groovy/org/codenarc/rule/unnecessary/UnnecessaryPublicModifierRule.groovy	(date 1554416536324)
@@ -80,6 +80,9 @@
         while (current <= node.lastLineNumber) {
             def line = sourceCode.line(current)
             if (line == null) {
+                if (node.isAbstract()) {
+                    return acc // can happen with abstract method in Parrot parser
+                }
                 LOG.warn("${rule.name} cannot find source code line $current in ${sourceCode.name}. Scanning lines ${node.lineNumber} to ${node.lastLineNumber}.")
                 return ''
             } else if (line.contains('{')) {
Index: src/main/java/org/codenarc/util/AstUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/main/java/org/codenarc/util/AstUtil.java	(revision 9a7cec02cb8cbaf845030f2434309e8968f5d7a7)
+++ src/main/java/org/codenarc/util/AstUtil.java	(date 1554377573185)
@@ -1197,7 +1197,7 @@
 
             // extract the relevant part of the last line
             if (lineIndex == node.getLastLineNumber() - 1) {
-                int stopIndex = node.getLastColumnNumber() < line.length() ? node.getLastColumnNumber() - 2 : line.length();
+                int stopIndex = node.getLastColumnNumber() < line.length() ? node.getLastColumnNumber() - 1 : line.length();
                 line = line.substring(0, stopIndex);
             }
 
```

**Comment by chrismair** [comment]
Thanks Paul. Yeah, the line and column number handling is one of the more complex, and likely brittle, parts of the CodeNarc code.

**Comment by chrismair** [comment]
Fixed by #423.
